File: Data Structures & Algorithms/task-scheduling/submission-1.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        freq={}
        for task in tasks:
            if task in freq:
                freq[task]+=1
            else:
                freq[task]=1
        maxFreq=max(freq.values())

        numMaxFreqTasks=0
        for count in freq.values():
            if count==maxFreq:
                numMaxFreqTasks+=1
        return max(
            len(tasks),
            (maxFreq-1)*(n+1)+numMaxFreqTasks
        )

        
        

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
         int l=0,r=0;
        int n=s.size();
        vector<int> count(26,0);
        int curr_max=0;
        int max_freq=0;
        while(r<n){
            int window_len=(r-l+1);
            int potential_max_freq=max(max_freq,count[s[r]-'A']+1);
            bool can_grow=(window_len-potential_max_freq<=k);
            if(can_grow){
                count[s[r]-'A']++;
                max_freq = max(max_freq, count[s[r] - 'A']);
                r+=1;
                curr_max=max(curr_max,r-l);
            }else if(l==r){
                l+=1;
                r+=1;
            }else{
                count[s[l]-'A']--;
                l+=1;
            }
        }
        return curr_max;
    }
};
